go/tools/bzltestutil/bincov/exit_hook.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2022 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bincov
+
+import (
+	"bytes"
+	"cmd/internal/cov"
+	"fmt"
+	"internal/runtime/exithook"
+	"log"
+	"os"
+	"runtime/coverage"
+
+	"github.com/bazelbuild/rules_go/go/tools/bzltestutil"
+)
+
+// Lock in the COVERAGE_DIR during test setup in case the test uses e.g. os.Clearenv.
+var coverageDir = os.Getenv("COVERAGE_DIR")
+
+func AddExitHook() {
+	if coverageDir == "" {
+		log.Printf("Not collecting coverage: COVERAGE_DIR is not set")
+		return
+	}
+
+	os.Setenv("GOCOVERDIR", coverageDir)
+
+	exithook.Add(exithook.Hook{
+		F: func() {
+			if err := coverage.WriteMetaDir(coverageDir); err != nil {
+				fmt.Fprintf(os.Stderr, "write meta: %v\n", err)
+				return
+			}
+			if err := coverage.WriteCountersDir(coverageDir); err != nil {
```

Comment:
Is this safe even if this binary is run multiple times concurrently? The metadata file apparently uses a hash of the file and is static, but I'm not completely sure about that either.

Reply:
Yeah, great question. [These writes are atomic](https://github.com/golang/go/blob/2a71af11fc7e45903be9ab84e59c668bb051f528/src/internal/coverage/cfile/emit.go#L41-L58)

Reply:
The files look like so:
```
covcounters.018e32252cedf32b4170bdfa58cdf42b.60747.1759602683829898000 - covmeta.018e32252cedf32b4170bdfa58cdf42b
```

i.e. counters have the pid and the timestamp in there, and are also [written atomically](https://github.com/golang/go/blob/2a71af11fc7e45903be9ab84e59c668bb051f528/src/internal/coverage/cfile/emit.go#L325-L329) so the case of a collision is a bit of data loss vs corruption. Seems fine.

Reply:
Is it obvious that the coverage reader creates below only reads back the files we add here vs. counter files corresponding to a different execution of the same binary, which may have different coverage?

In any case, could you add a comment here describing this and the other relevant properties (atomic writes, globally unique or content described names) and why they hold? This seems pretty subtle and if we get it wrong, coverage may be non-deterministic.

We could also emit the files into a unique temp subdirectory instead, which I think is obviously safe.

Reply:
yeah I also considered the unique tempdir and hesitated, but let's just do it. it's more obvious

Reply:
(done)